arrayLib.js

```diff
@@ -0,0 +1,258 @@
+		}
+
+		this.array = newArr;
+
+		return (arguments.length == 1) ? this : newArr;
+	}
+
+
+	function mapArray() {
+		var array, callback;
+
+		if (arguments.length == 1) {
+			array = this.array;
+			callback = arguments[0];
+		}
+
+		if (arguments.length == 2) {
+			array = arguments[0];
+			callback = arguments[1];
+		}
+
+		if (!arrayCheck(array) || !functionCheck(callback)) {
+			console.log('error');
+			return;
+		}
+
+		var newArr = [];
+
+		for (var i = 0; i < array.length; i++) {
+			var newValue = callback(array[i]);
+			newArr.push(newValue);
+		}
+
+		this.array = newArr;
+
+		return (arguments.length == 1) ? this : newArr;
+	}
+
+
+	function reduceArray() {
+		var array, callback, initialValue;
+
+		switch (arguments.length) {
+			case 1: array = this.array;
+				callback = arguments[0];
+				initialValue = 0;
+				break;
+			case 2: array = arguments[0];
+				callback = arguments[1];
+				initialValue = 0;
+				break;
+			case 3: array = arguments[0];
+				callback = arguments[1];
+				initialValue = arguments[2];
+				break;
+		}
+
+		if (!arrayCheck(array) || !functionCheck(callback)) {
+			console.log('error');
+			return;
+		}
+
+		for (var i = 0; i < array.length; i++) {
+			initialValue = callback(initialValue, array[i]);
+		}
+
+		return initialValue;
+	}
+
+
+	function filterArray() {
+		var array, callback;
+
+		if (arguments.length == 1) {
+			array = this.array;
+			callback = arguments[0];
+		}
+
+		if (arguments.length == 2) {
+			array = arguments[0];
+			callback = arguments[1];
+		}
+
+		if (!arrayCheck(array) || !functionCheck(callback)) {
+			console.log('error');
+			return;
+		}
+		var newArr = [];
+
+		for (var i = 0; i < array.length; i++) {
+			if (callback(array[i])) {
+				newArr.push(array[i]);
+			}
+		}
+
+		this.array = newArr;
+
+		return (arguments.length == 1) ? this : newArr;
+	}
+
+
+	function arrayForEach() {
+		var array, callback;
+
+		if (arguments.length == 1) {
+			array = this.array;
+			callback = arguments[0];
+		}
+
+		if (arguments.length == 2) {
+			array = arguments[0];
+			callback = arguments[1];
+		}
+
+		if (!arrayCheck(array) || !functionCheck(callback)) {
+			console.log('error');
+			return;
+		}
+
+		for (var i = 0; i < array.length; i++) {
+			callback(array[i]);
+		}
+
+		return (arguments.length == 1) ? this : null;
+	}
+
+
+	function value() {
+		return this.array || [];
+	}
+
+
+	function arrayCheck(array) {
+		if (array.constructor != Array || array.length == 0) {
+			return false;
+		}
+		return true;
+	}
+
+	function lengthCheck(array, n) {
+		if (array.length != undefined && n > array.length) {
+			return false;
+		}
+		return true;
+	}
+
+	function functionCheck(func) {
+		if (func.constructor != Function) {
+			return false;
+		}
+		return true;
+	}
+
+	return {
+		chain: chain,
+		take: takeNArrayVal,
+		skip: skipNArrayVal,
+		map: mapArray,
+		reduce: reduceArray,
+		filter: filterArray,
+		forEach: arrayForEach,
+		value: value
+	};
+})();
```

Comment:
You can use another approach to return object with needed properties. You can declare this in the beginning of the function and apply function to it and in the end return this.

function module(){
    let this = this || {};

    this.reduce = function(){};

    return this;
}()